### ISSUE TITLE
AA only affects World

### ISSUE THREAD

**Issue by lafoxxx92** [OldUnreal/UnrealTournamentPatches#606]
![image](https://user-images.githubusercontent.com/34402711/143781693-f20fa919-61aa-4355-8ad8-27e6e3706b61.png)

**Comment by CacoFFF** [comment]
Disable NoFiltering

**Comment by CacoFFF** [comment]
After reading https://github.com/OldUnreal/UnrealTournamentPatches/issues/609, I'm getting the idea that what you mean is to have pixelated graphics, but with the color transitions being anti-aliased right?

Native antialiasing implementations only run multi-sampling on geometry edges, in order to achieve actual in-texture anti-aliasing one of two things need to happen:
1- Render the game at 2x the resolution and downscale.
2- Implement a specific shader capable of interpolating textures at screen pixel range.

I've made some shader experiments with option 2, one resulting in masked textures being anti-aliased at their edges, but the effect is not precise and doesn't interact well with Level-of-Detail options.
Will try again after the OpenGLDrv's modern shader system is fully implemented*

* OpenGLDrv uses legacy OpenGL 1/2 contexts with ARB assemby shaders, but it's planned to add capabilities to support many OpenGL implementations and their shading versions which should vastly improve performance and capabilities in the future.

**Comment by lafoxxx92** [comment]
> what you mean is to have pixelated graphics, but with the color transitions being anti-aliased right?

Yes. I want to have all polygons (world and models) antialiased (it currently works, see the edge of the wall), but the textures should look pixelated.
Disabling NoFilter doesn't help, only Textures get smooth, but Model Polygons are still jagged (see the edge of the rocket launcher --it's still pixelated, even when you turn off NoFilter it doesn't get antialiased like the wall edge).

**Comment by Anxiett** [comment]
MSAA only filters polygons, MSAA is also applied to the rocket launcher:

https://i.imgur.com/uUpjsVq.png

Most of the edges are simply drawn into the textures, that's why they appear like AA isn't applied there. FSAA most likely can filter these out, but it may slightly blur a few things too, like far away textures and it maybe not entirely will give you the wanted results. Not sure if FSAA can be applied outside the driver, but it might be possible to add it.

Personally i think adding the detail layers is a good way to get rid off the bluriness the textures have when they are filtered.

ioquake3 uses an opengl renderer called opengl2 https://github.com/ioquake/ioq3/blob/main/opengl2-readme.md

It adds an upsampling method, which isn't as sharp as waifu, but not so far away easier at 2x, its much better than linear filtering and has no artefacts like all AI upscalers do.
Would be great to have it in UT too, not sure if it will work good with detail layers, but maybe works great too.

Pixel look is something a game should be designed for, like the build engine games, perhaps quake 1 too as it came shortly before voodoo came out, but beyond i think filtered textures is pretty much the look designers were working with.

**Comment by CacoFFF** [comment]
I've been doing experiments with Intel's CMAA using their native extension, unfortunately this only works with Intel GPU's but I opted against including this in v469e due to various issues with it in the Intel driver.

https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/issues/877

I may try to implement CMAA2 in the future for all GPU's but that's a considerable amount of work.
